### PR TITLE
[ci] window jobs based on the platform

### LINKED ISSUE #9213: Windows CI: source cache is not reused between CI runs
This is similar to #9212, but is a about the cache for downloaded source archives. This might be more controversial, because it might not be easy to setup a source cache such that malicious pull requests cannot badly influence the outcome of future runs on the same machine. One can easily achieve that existing files cannot be overwritten, but it would be tricky to ensure that only properly named files are in the cache.

### PR DESCRIPTION
**Pull request by gares**
This PR builds a minimal platform on windows 64 and 32 bits (just Coq and CoqIDE).
We use the master branch of the platform to fetch the scripts, the URL is hardcoded in the .gitlab-ci.yml file.

Closes #9213

### CODE REVIEW

File: dev/ci/platform-windows.bat
Comment (MSoegtropIMC) [inline dev/ci/platform-windows.bat:22]: This is no longer used for the platform - the dest folder is ~/.opam

File: dev/ci/platform-windows.bat
Comment (MSoegtropIMC) [inline dev/ci/platform-windows.bat:26]: This is no longer used for the platform - the dest folder is ~/.opam

File: dev/ci/platform-windows.bat
Comment (MSoegtropIMC) [inline dev/ci/platform-windows.bat:47]: -compcert and -vst are not required if -extent=i, but don't hurt either

File: dev/ci/platform-windows.bat
Comment (MSoegtropIMC) [inline dev/ci/platform-windows.bat:42]: That's the default and given in the example file only for illustration. I think for CI we should keep the default, so that it changes in CI if the default changes.

**Review by MSoegtropIMC — COMMENTED** [review]
LGTM - if you need help to get it work please ping me.

File: dev/ci/platform-windows.bat
Comment (gares) [reply dev/ci/platform-windows.bat:28]: Hum, my understanding was that this was needed in order to not get clash between jobs.

File: dev/ci/platform-windows.bat
Comment (gares) [reply dev/ci/platform-windows.bat:49]: ok

File: dev/ci/platform-windows.bat
Comment (gares) [reply dev/ci/platform-windows.bat:44]: ok

**Comment by MSoegtropIMC** [comment]
One related question: do you think I should setup and keep the master branch of the Coq platform so that it builds master versions of all packages? Would this be better done by local patch packages or by including the dev repo?

**Comment by gares** [comment]
I'd like some branch, maybe master, to be tracked by this CI job. It only needs Coq and CoqIDE.

Then, in a CI job of the platform (not of Coq) I think it may be worth generating an overlay of packages built out of Coq's ci-basic-overlay.sh, and build that once a day. I think Coq can test extra packages on linux only.

**Comment by gares** [comment]
Any idea what I do wrong? I run out of ideas about "downloading a file"...

**Comment by Zimmi48** [comment]
> I'd like some branch, maybe master, to be tracked by this CI job. It only needs Coq and CoqIDE.

Yes, to merge this in `master`, it would be best to base this branch on the `master` branch and not the `v8.13` branch of the platform.

Also, could you remove the `dev/build/windows` directory? I will be relieved that we don't have to maintain these scripts anymore.

**Comment by Zimmi48** [comment]
> Any idea what I do wrong? I run out of ideas about "downloading a file"...

IMHO the issue is that the file that you downloaded is not named "platform.zip".

**Comment by ejgallego** [comment]
May be a very naïve question [sorry if it is], but wasn't the goal of the platform to actually have it perform the builds in their own CI, as to reduce Coq's CI windows build to a regular build?

**Comment by Zimmi48** [comment]
Yes, but this is precisely what is happening in this PR (we reduce the testing on the Coq side a lot).

**Comment by gares** [comment]
I did not touch the azure job for now. There is a lot of duplication (setting up cygwin) but that job also runs dune to build Coq, which is not what we do when we use the opam package. If we converge on that front, then the azure job becomes useless and can be remove as well (we can still use azure, but the job should just call the bat script in this PR)

**Comment by Zimmi48** [comment]
The fact that the Azure job calls Dune is not important (it is sufficiently tested). It would be very good to stop relying on GitLab CI for Windows testing, so yes, I'd be happy to have this job moved to Azure (and the old Azure job removed).

**Comment by gares** [comment]
I've to pass the ball now. @MSoegtropIMC there is a problem with depext I don't understand and I have to move to something else. Feel free to push to this branch (even force push).

**Comment by ejgallego** [comment]
Indeed if the platform is opam-based, our opam testing should suffice, and @gares also conveniently added the test for the relocation feature to the test suite.

**Comment by gares** [comment]
@ejgallego I pushed an extra commit to make it clear to you what we are doing here

**Comment by gares** [comment]
The plan is the following one, I believe.

This Coq CI job is to test that:
- at release time, one can build a minimal windows installer (coq + coqide)
- at development time we don't break the scripts of the platform

Other Coq CI jobs test that:
- we have a commit that works for all plugins and some libraries included in the platform (on linux only)

The CI of the platform tests that:
- all plugins and all libraries part of the platform build (peeking the overlay hash from Coq, for the ones checked by Coq)

**Comment by MSoegtropIMC** [comment]
@gares : sorry I am busy until the evening (6 PM or so). I will have a look then.

I would suggest that we create a coq_ci branch which is dedicated to the needs of Coq's CI - e.g. with a right away reduced package file and maybe a local opam file for coq and coqide master).

@ejgallego : the idea is to reuse the opam based build scripts from the platform to just build Coq and CoqIDE on Windows. Building CoqIDE on Windows is not trivial, so it makes sense to reuse the opam based build scripts from the platform for this. One could do the same for Mac, btw (although there is no Mac installer build script as yet, comming soon).

**Comment by gares** [comment]
> I would suggest that we create a coq_ci branch which is dedicated to the needs of Coq's CI - e.g. with a right away reduced package file and maybe a local opam file for coq and coqide master).

fine for me, and no hurry.

**Comment by Zimmi48** [comment]
> > I would suggest that we create a coq_ci branch which is dedicated to the needs of Coq's CI - e.g. with a right away reduced package file and maybe a local opam file for coq and coqide master).
> 
> fine for me, and no hurry.

Fine for me, but I'm worried about the multiplication of branches.

**Comment by gares** [comment]
Given the nature of the platform, I would probably use master (and the option `-extent=i` in Coq's CI).

If we want to generate the overlay packages on the fly, for the platform CI, I'd rather keep all packages there, both in the list and the overlay repo, with a `.dev` version, as "templates" to be filled in by the CI script with the right `url#$hash` metadata. Then, at release time one branches from master vX.Y and starts to pin hashes. A bit like we do in Coq with the script https://github.com/coq/coq/blob/master/dev/tools/pin-ci.sh

**Comment by MSoegtropIMC** [comment]
Maybe we can discuss this in the next Coq call.

**Comment by gares** [comment]
Sure, but the next one it is already loaded of stuff: https://github.com/coq/coq/wiki/Coq-Call-2020-12-09

**Comment by MSoegtropIMC** [comment]
@gares : since there seems to be going on something strange with depext, I did a test run of the coq platform script alone on one of the runners - did work fine. I will have a look at running your script tomorrow.

**Comment by MSoegtropIMC** [comment]
@gares: just to avoid double work: I am currently testing your CI wrapper script interactively on one of the runners (the trusted one which is normally not used).

**Comment by gares** [comment]
I'm not working on this today

**Comment by MSoegtropIMC** [comment]
It still fails here:
```
<><> Carrying on to "C:\ci\cygwin64_11095_13585\usr\x86_64-w64-mingw32\sys-root\mingw\bin\opam.exe depext conf-gtksourceview3" 
# Detecting depexts using vars: arch=x86_64, os=win32, os-distribution=cygwinports, os-family=windows
# The following system packages are needed:
Fatal error: exception Sys_error("Bad file descriptor")
```
Running this locally on the runner works. I guess I do wrong assumptions about the environment. I will add a SET command at the beginning and compare the local and the CI build output.

**Comment by Zimmi48** [comment]
When will this be finished? If not soon, maybe we should try to fix (or at least disable) the currently failing Windows jobs.

**Comment by MSoegtropIMC** [comment]
Should be tomorrow or Wednesday.

**Comment by gares** [comment]
This PR is finally ready for review/merge.

**Comment by gares** [comment]
The only open question to me is: should we continue use Inria's windows workers via gitlab or just use GH the actions ones?

**Comment by gares** [comment]
CI green

**Comment by SkySkimmer** [comment]
>The only open question to me is: should we continue to use Inria's Windows workers via gitlab, or switch to the ones available via github actions?

So long as they're beefy enough the github ones should need less maintenance.
Do we know how many workers we get on github actions?

File: .gitlab-ci.yml
Comment (SkySkimmer) [inline .gitlab-ci.yml:353]: Why's this not in the template?

**Comment by Zimmi48** [comment]
> The only open question to me is: should we continue to use Inria's Windows workers via gitlab, or switch to the ones available via github actions?

I'd like to switch (to reduce the maintenance burden) but this should probably be part of another PR.

File: .gitlab-ci.yml
Comment (SkySkimmer) [inline .gitlab-ci.yml:265]: Incorrect change afaict, if you click on download eg at https://gitlab.com/coq/coq/-/jobs/942269583 the filename is `_CI_JOB_NAME.zip`

File: .gitlab-ci.yml
Comment (SkySkimmer) [inline .gitlab-ci.yml:361]: Is the idea that since we build less we can now build more?

File: .gitlab-ci.yml
Comment (gares) [reply .gitlab-ci.yml:353]: The template is for testing the platform. It is now only run on windows, but we should also run a quick job on osx and linux.

If we move to GH actions then this section goes away anyway... if not, I'd like to keep things as they are (unless one can do multiple inheritance)

File: .gitlab-ci.yml
Comment (gares) [reply .gitlab-ci.yml:361]: I've added ti back (I don't know what it does, not why I did remove it)

**Comment by gares** [comment]
github actions are quite generous: https://docs.github.com/en/free-pro-team@latest/actions/reference/usage-limits-billing-and-administration (20 concurrent jobs, 5 macosx ones)

File: .gitlab-ci.yml
Comment (Zimmi48) [inline .gitlab-ci.yml:355]: You don't want any platform test on pull requests?

**Review by Zimmi48 — COMMENTED** [review]
This line should go away now:
https://github.com/coq/coq/blob/a22da3e70551658deefbbedf261acdc3ead5403d/.github/CODEOWNERS#L19

File: .gitlab-ci.yml
Comment (gares) [reply .gitlab-ci.yml:355]: I thought that @SkySkimmer suggested this

File: .gitlab-ci.yml
Comment (Zimmi48) [reply .gitlab-ci.yml:361]: It excludes the Windows 32 bit job from PR builds because it's usually sufficient to test one architecture.

**Review by Zimmi48 — APPROVED** [review]
The one thing I didn't review is the `.bat` file.

File: .gitlab-ci.yml
Comment (gares) [reply .gitlab-ci.yml:361]: Hum, I'm fine with that, but then I'll remove the entire job.

Comment (Zimmi48) [reply .gitlab-ci.yml:361]: Yes, that makes sense if you're already testing this with a scheduled job on the platform repo.

Comment (gares) [reply .gitlab-ci.yml:361]: done

**Comment by gares** [comment]
This needs an asignee

**Comment by Zimmi48** [comment]
AFAIU this PR is ready. @MSoegtropIMC: did you forget about it?

**Comment by MSoegtropIMC** [comment]
@Zimmi48 : no, I was just a bit busy after my vacation. I planned this for today evening.

**Comment by Zimmi48** [comment]
Perfect, thanks!

**Pull request merged**
